File: apps/openwebui/backend/open_webui/routers/benchmarks/profiles.py
"""routers/benchmarks/profiles.py - CRUD for stored serving profiles.

Backs decision 11 in docs/migration-plan.md: full profile CRUD (list,
create, clone, edit, archive, set default) from the Serve page. Thin over
models/benchmark_profiles.py.BenchmarkProfileTable, which already carries
the actual rules (versioning, the default-profile guard rails); this file's
only job is turning those into HTTP.

A profile's `name` is immutable after creation (tuning search spaces are
files named after it -- see the plan's "Assumptions requiring
confirmation"). `DefinitionForm` has no `name` field at all, and every form
here is `extra='forbid'`, so a client that tries to smuggle a rename through
an edit gets a loud 422 rather than a silently ignored field.
"""

from __future__ import annotations

from fastapi import APIRouter, Depends, HTTPException, status
from open_webui.benchmarks.serving.profiles import ProfileError
from open_webui.constants import ERROR_MESSAGES
from open_webui.models.benchmark_profiles import BenchmarkProfileEntry, BenchmarkProfiles
from open_webui.utils.auth import get_admin_user
from pydantic import BaseModel, ConfigDict

router = APIRouter()


class DefinitionForm(BaseModel):
    """Everything a profile *version* carries.

    Deliberately excludes `name`, `display_name` and `is_default`: those
    belong to the profile's identity (`BenchmarkProfile`), not its
    definition, and are never accepted here -- see this module's docstring.
    """

    arch: str
    alias: str
    model_path: str
    hf_repo: str = ''
    hf_pattern: str = ''
    ctx: int
    threads: int
    ngl: int
    moe: int | None = None
    override_tensors: str | None = None
    parallel: int = 1
    cache_k: str = 'q8_0'
    cache_v: str = 'q8_0'
    batch: int = 512
    ubatch: int = 512
    spec: list[str] = []
    samplers: list[str] = []
    extra: list[str] = []
    reasoning_effort_default: str | None = None
    notes: str = ''

    model_config = ConfigDict(extra='forbid')


class CreateProfileForm(BaseModel):
    name: str
    display_name: str
    definition: DefinitionForm
    note: str | None = None

    model_config = ConfigDict(extra='forbid')


class CloneProfileForm(BaseModel):
    """A new identity for an existing profile's *current* definition.

    The clone's definition is read from the source profile's latest version
    server-side, not accepted from the client -- a clone is meant to start
    identical, with `name`/`display_name` the only things that differ.
    """

    name: str
    display_name: str
    note: str | None = None

    model_config = ConfigDict(extra='forbid')


class AddVersionForm(BaseModel):
    definition: DefinitionForm
    note: str | None = None

    model_config = ConfigDict(extra='forbid')


class DisplayNameForm(BaseModel):
    display_name: str

    model_config = ConfigDict(extra='forbid')


#: Fields BenchmarkProfileVersionModel carries beyond DEFINITION_FIELDS --
#: dropped when cloning a version into a new profile's own first version.
_VERSION_IDENTITY_FIELDS = {'version_id', 'profile_id', 'version', 'created_at', 'created_by', 'note'}


async def _get_or_404(name: str) -> BenchmarkProfileEntry:
    entry = await BenchmarkProfiles.get_by_name(name)
    if entry is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=ERROR_MESSAGES.NOT_FOUND)
    return entry


@router.get('/')
async def list_profiles(include_archived: bool = False, user=Depends(get_admin_user)):
    return await BenchmarkProfiles.list_profiles(include_archived=include_archived)


@router.get('/default')
async def get_default_profile(user=Depends(get_admin_user)):
    entry = await BenchmarkProfiles.get_default()
    if entry is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=ERROR_MESSAGES.NOT_FOUND)
    return entry


@router.get('/{name}')
async def get_profile(name: str, user=Depends(get_admin_user)):
    return await _get_or_404(name)


@router.get('/{name}/versions')
async def list_versions(name: str, user=Depends(get_admin_user)):
    entry = await _get_or_404(name)
    return await BenchmarkProfiles.list_versions(entry.profile.profile_id)


@router.post('/', status_code=status.HTTP_201_CREATED)
async def create_profile(form_data: CreateProfileForm, user=Depends(get_admin_user)):
    if await BenchmarkProfiles.get_by_name(form_data.name) is not None:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail=ERROR_MESSAGES.DEFAULT(f"a profile named '{form_data.name}' already exists"),
        )
    try:
        return await BenchmarkProfiles.create(
            name=form_data.name,
            display_name=form_data.display_name,
            definition=form_data.definition.model_dump(),
            created_by=user.id,
            note=form_data.note,
        )
    except ProfileError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e)) from e


@router.post('/{name}/clone', status_code=status.HTTP_201_CREATED)
async def clone_profile(name: str, form_data: CloneProfileForm, user=Depends(get_admin_user)):
    source = await _get_or_404(name)
    if await BenchmarkProfiles.get_by_name(form_data.name) is not None:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail=ERROR_MESSAGES.DEFAULT(f"a profile named '{form_data.name}' already exists"),
        )
    definition = source.version.model_dump(exclude=_VERSION_IDENTITY_FIELDS)
    try:
        return await BenchmarkProfiles.create(
            name=form_data.name,
            display_name=form_data.display_name,
            definition=definition,
            created_by=user.id,
            note=form_data.note or f"cloned from '{name}' (version {source.version.version})",
        )
    except ProfileError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e)) from e


@router.post('/{name}/versions', status_code=status.HTTP_201_CREATED)
async def add_version(name: str, form_data: AddVersionForm, user=Depends(get_admin_user)):
    entry = await _get_or_404(name)
    try:
        return await BenchmarkProfiles.add_version(
            entry.profile.profile_id,
            definition=form_data.definition.model_dump(),
            created_by=user.id,
            note=form_data.note,
        )
    except ProfileError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e)) from e


@router.post('/{name}/display-name')
async def set_display_name(name: str, form_data: DisplayNameForm, user=Depends(get_admin_user)):
    entry = await _get_or_404(name)
    await BenchmarkProfiles.set_display_name(entry.profile.profile_id, form_data.display_name)
    return await _get_or_404(name)


@router.post('/{name}/default')
async def set_default(name: str, user=Depends(get_admin_user)):
    entry = await _get_or_404(name)
    try:
        await BenchmarkProfiles.set_default(entry.profile.profile_id)
    except ProfileError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e)) from e
    return await _get_or_404(name)


@router.post('/{name}/archive')
async def archive_profile(name: str, user=Depends(get_admin_user)):
    entry = await _get_or_404(name)
    try:
        await BenchmarkProfiles.archive(entry.profile.profile_id)
    except ProfileError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e)) from e
    return await _get_or_404(name)


@router.post('/{name}/unarchive')
async def unarchive_profile(name: str, user=Depends(get_admin_user)):
    entry = await _get_or_404(name)
    await BenchmarkProfiles.unarchive(entry.profile.profile_id)
    return await _get_or_404(name)

File: apps/web/src/lib/apis/benchmarks/profiles.ts
// New code, not a port: backend/open_webui/routers/benchmarks/profiles.py's
// CRUD endpoints (list, get, get-default, list-versions, create, clone,
// add-version, set-display-name, set-default, archive, unarchive) have no
// SvelteKit frontend at all -- see this file's own docstring, and decision
// 11 in docs/migration-plan.md ("Full CRUD from the Serve page"), which the
// existing Serve.svelte never actually built. Follows the same fetch/error
// shape as the rest of apps/openwebui/src/lib/apis/benchmarks/index.ts
// (ported verbatim elsewhere) so this reads as one module, not two styles.

import { WEBUI_API_BASE_URL } from '@/lib/constants';

const BASE = `${WEBUI_API_BASE_URL}/benchmarks/profiles`;

const request = async (path: string, token: string, init?: RequestInit) => {
	let error: unknown = null;

	const res = await fetch(`${BASE}${path}`, {
		...init,
		headers: {
			Accept: 'application/json',
			'Content-Type': 'application/json',
			authorization: `Bearer ${token}`,
			...init?.headers
		}
	})
		.then(async (res) => {
			if (!res.ok) throw await res.json();
			return res.json();
		})
		.catch((err) => {
			error = err?.detail ?? err;
			console.error(err);
			return null;
		});

	if (error) {
		throw error;
	}

	return res;
};

export type ProfileDefinition = {
	arch: string;
	alias: string;
	model_path: string;
	hf_repo?: string;
	hf_pattern?: string;
	ctx: number;
	threads: number;
	ngl: number;
	moe?: number | null;
	override_tensors?: string | null;
	parallel?: number;
	cache_k?: string;
	cache_v?: string;
	batch?: number;
	ubatch?: number;
	spec?: string[];
	samplers?: string[];
	extra?: string[];
	reasoning_effort_default?: string | null;
	notes?: string;
};

export type ProfileModel = {
	profile_id: number;
	name: string;
	display_name: string;
	is_default: boolean;
	created_at: number;
	archived_at: number | null;
};

export type ProfileVersionModel = ProfileDefinition & {
	version_id: number;
	profile_id: number;
	version: number;
	created_at: number;
	created_by: string | null;
	note: string | null;
};

export type ProfileEntry = {
	profile: ProfileModel;
	version: ProfileVersionModel;
};

export const listProfiles = async (
	token: string = '',
	includeArchived: boolean = false
): Promise<ProfileEntry[]> =>
	request(`/?include_archived=${includeArchived}`, token, { method: 'GET' });

export const getProfile = async (token: string = '', name: string): Promise<ProfileEntry> =>
	request(`/${encodeURIComponent(name)}`, token, { method: 'GET' });

export const listProfileVersions = async (
	token: string = '',
	name: string
): Promise<ProfileVersionModel[]> =>
	request(`/${encodeURIComponent(name)}/versions`, token, { method: 'GET' });

export const createProfile = async (
	token: string = '',
	form: { name: string; display_name: string; definition: ProfileDefinition; note?: string }
): Promise<ProfileEntry> =>
	request('/', token, { method: 'POST', body: JSON.stringify(form) });

export const cloneProfile = async (
	token: string = '',
	name: string,
	form: { name: string; display_name: string; note?: string }
): Promise<ProfileEntry> =>
	request(`/${encodeURIComponent(name)}/clone`, token, {
		method: 'POST',
		body: JSON.stringify(form)
	});

export const addProfileVersion = async (
	token: string = '',
	name: string,
	form: { definition: ProfileDefinition; note?: string }
): Promise<ProfileEntry> =>
	request(`/${encodeURIComponent(name)}/versions`, token, {
		method: 'POST',
		body: JSON.stringify(form)
	});

export const setProfileDisplayName = async (
	token: string = '',
	name: string,
	displayName: string
): Promise<ProfileEntry> =>
	request(`/${encodeURIComponent(name)}/display-name`, token, {
		method: 'POST',
		body: JSON.stringify({ display_name: displayName })
	});

export const setDefaultProfile = async (token: string = '', name: string): Promise<ProfileEntry> =>
	request(`/${encodeURIComponent(name)}/default`, token, { method: 'POST' });

export const archiveProfile = async (token: string = '', name: string): Promise<ProfileEntry> =>
	request(`/${encodeURIComponent(name)}/archive`, token, { method: 'POST' });

export const unarchiveProfile = async (token: string = '', name: string): Promise<ProfileEntry> =>
	request(`/${encodeURIComponent(name)}/unarchive`, token, { method: 'POST' });
